text/0000-unicode-idents.md

```diff
@@ -0,0 +1,162 @@
+
+It always a possibility to do nothing and limit identifiers to ASCII.
+
+It has been suggested that Unicode identifiers should be opt-in instead of opt-out. The proposal chooses opt-out to benefit the international Rust community. New Rust users should not need to search for the configuration option they may not even know exists. Additionally it simplifies tutorials in other languages as they can omit an annotation in every code snippet.
+
+## Confusable detection
+
+The current design was chosen because the algorithm and list of similar characters are already provided by the Unicode Consortium. A different algorithm and list of characters could be created. I am not aware of any other programming language implementing confusable detection. The confusable detection was primarily included because homoglyph attacks are a huge concern for some member of the community.
+
+Instead of offering confusable detection the lint `forbid(unicode_idents)` is sufficient to protect project written in English from homoglyph attacks. Projects using different languages are probably either written by students, by a small group or inside a regional company. These projects are not threatened as much as large open source projects by homoglyph attacks but still benefit from the easier debugging of typos.
+
+# Prior art
+[prior-art]: #prior-art
+
+"[Python PEP 3131][PEP 3131]: Supporting Non-ASCII Identifiers" is the Python equivalent to this proposal. The proposed identifier grammar **XID_Start&nbsp;XID_Continue<sup>\*</sup>** is identical to the one used in Python 3.
+
+[JavaScript] supports Unicode identifiers based on the same Default Identifier Syntax but does not apply normalization.
+
+The [CPP reference][C++] describes the allowed Unicode identifiers it is based on the immutable identifier principle.
+
+[Java] also supports Unicode identifiers. Character must belong to a number of Unicode character classes similar to XID_start and XID_continue used in Python. Unlike in Python no normalization is performed.
+
+The [Go language][Go] allows identifiers in the form **Letter (Letter | Number)\*** where **Letter** is a Unicode letter and **Number** is a Unicode decimal number. This is more restricted than the proposed design mainly as is does not allow combining characters needed to write some languages such as Hindi.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Which context is adequate for confusable detection: file, current scope, crate?
+* Are Unicode characters allowed in `no_mangle` and `extern fn`s?
+* How do Unicode names interact with the file system?
+* Are crates with Unicode names allowed and can they be published to crates.io?
+* Are `unicode_idents` and `confusable_unicode_idents` good names?
+* Should [ZWNJ and ZWJ be allowed in identifiers][TR31Layout]?
+* Should *rustc* accept files in a different encoding than *UTF-8*?
```

Comment:
This one is easy: no. Why would it? UTF-8 by definition supports all of Unicode. Also, source files encodings seem off-topic for this RFC.